renderer/backend/RenderBackend.cpp

```diff
@@ -0,0 +1,256 @@
+
+******************************************************************************/
+
+#include "precompiled.h"
+#pragma hdrstop
+
+#include "RenderBackend.h"
+
+#include "../AmbientOcclusionStage.h"
+#include "../Profiling.h"
+#include "../GLSLProgram.h"
+#include "../FrameBufferManager.h"
+
+RenderBackend renderBackendImpl;
+RenderBackend *renderBackend = &renderBackendImpl;
+
+idCVar r_useNewBackend( "r_useNewBackend", "0", CVAR_BOOL|CVAR_RENDERER|CVAR_ARCHIVE, "Use experimental new backend" );
+idCVar r_useBindlessTextures("r_useBindlessTextures", "1", CVAR_BOOL|CVAR_RENDERER|CVAR_ARCHIVE, "Use experimental bindless texturing to reduce drawcall overhead (if supported by hardware)");
+
+RenderBackend::RenderBackend() 
+	: depthStage( &drawBatchExecutor ),
+	  interactionStage( &drawBatchExecutor ),
+	  stencilShadowStage( &drawBatchExecutor )
+{}
+
+void RenderBackend::Init() {
+	drawBatchExecutor.Init();
+	depthStage.Init();
+	interactionStage.Init();
+	stencilShadowStage.Init();
+}
+
+void RenderBackend::Shutdown() {
+	stencilShadowStage.Shutdown();
+	interactionStage.Shutdown();
+	depthStage.Shutdown();
+	drawBatchExecutor.Destroy();
+}
+
+void RenderBackend::DrawView( const viewDef_t *viewDef ) {
+	// we will need to do a new copyTexSubImage of the screen when a SS_POST_PROCESS material is used
+	backEnd.currentRenderCopied = false;
+	backEnd.afterFogRendered = false;
+
+	GL_PROFILE( "DrawView" );
+
+	// skip render bypasses everything that has models, assuming
+	// them to be 3D views, but leaves 2D rendering visible
+	if ( viewDef->viewEntitys && r_skipRender.GetBool() ) {
+		return;
+	}
+
+	// skip render context sets the wgl context to NULL,
+	// which should factor out the API cost, under the assumption
+	// that all gl calls just return if the context isn't valid
+	if ( viewDef->viewEntitys && r_skipRenderContext.GetBool() ) {
+		GLimp_DeactivateContext();
+	}
+	backEnd.pc.c_surfaces += viewDef->numDrawSurfs;
+
+	RB_ShowOverdraw();
+
+
+	int processed;
+
+	backEnd.depthFunc = GLS_DEPTHFUNC_EQUAL;
+
+	// clear the z buffer, set the projection matrix, etc
+	RB_BeginDrawingView();
+
+	backEnd.lightScale = r_lightScale.GetFloat();
+	backEnd.overBright = 1.0f;
+
+	drawSurf_t **drawSurfs = ( drawSurf_t ** )&viewDef->drawSurfs[ 0 ];
+	int numDrawSurfs = viewDef->numDrawSurfs;
+
+	// if we are just doing 2D rendering, no need to fill the depth buffer
+	if ( viewDef->viewEntitys ) {
+		// fill the depth buffer and clear color buffer to black except on subviews
+		depthStage.DrawDepth( viewDef, drawSurfs, numDrawSurfs );
+		if( ambientOcclusion->ShouldEnableForCurrentView() ) {
+			ambientOcclusion->ComputeSSAOFromDepth();
+		}
+		DrawShadowsAndInteractions( viewDef );
+	}
+		
+	// now draw any non-light dependent shading passes
+	int RB_STD_DrawShaderPasses( drawSurf_t **drawSurfs, int numDrawSurfs );
+	processed = RB_STD_DrawShaderPasses( drawSurfs, numDrawSurfs );
+
+	// fog and blend lights
+	void RB_STD_FogAllLights( bool translucent );
```

Comment:
I suggest marking it as extern.
An eye is not trained at all in seeing function declarations in function bodies =)